docs/proposals/dedicated-nodes.md

```diff
@@ -0,0 +1,210 @@
+of this proposal, taints are only associsted with nodes and tolerations are
+only associated with pods, but in theory each could also be associated with other
+types of resources. This document only describes how to use taints and tolerations
+to enforce dedicated nodes; in a followup document we will describe other anticipated uses.
+(Some are alluded to in #17190, the issue on which this proposal is based.)
+
+With taints and tolerations, implementing dedicated nodes is simple: every node that is
+part of a dedicated node group gets a taint of the form `dedicated=<group name>`, and
+every pod that is allowed to use a node in that group gets a toleration of the form
+`dedicated=<group name>`. The vast majority of pods, which are not in any dedicated node
+group, need not be touched, since the taint automatically prevents them from scheduling
+onto the dedicated nodes. This approach easily generalizes to pods that are allowed to
+schedule into multiple dedicated node groups, and nodes that are a member of multiple
+dedicated node groups.
+
+## API
+
+```go
+// The object this Taint is attached to has the effect "effect" on
+// any object that that does not tolerate the Taint.
+type Taint struct {
+	Key string  `json:"key" patchStrategy:"merge" patchMergeKey:"key"`
+	Value string  `json:"value,omitempty"`
+	Effect TaintEffect  `json:"effect"`
+}
+
+// The object this Toleration is attached to tolerates any taint
+// that matches the triple <key,value,effect>.
+type Toleration struct
+	Key string  `json:"key" patchStrategy:"merge" patchMergeKey:"key"`
+	// Wildcard (*) is allowed for value, so that a pod can tolerate all taints
+	// of a particular category. But value is otherwise just a regular string.
+	Value string  `json:"value,omitempty"`
+	Effect TaintEffect  `json:"effect"`
+}
+
+type TaintEffect string
+
+const (
+	// Do not allow new pods to start unless they tolerate the taint,
+	// and allow all already-running pods to continue running.
+	TaintEffectNoStart TaintEffect = "noStart"
+	// Do not allow new pods to start unless they tolerate the taint,
+	// and evict all already-running pods that do not tolerate the taint.
+	TaintEffectEvictAndNoStart TaintEffect = "evictAndNoStart"
+```
+
+We will add
+
+```go
+	Tolerations []Toleration  `json:"tolerations,omitempty"`
+```
+
+to `PodSpec`. A pod must tolerate all of a node's taints in order to be able
+to schedule onto that node.
+
+We will add
+
+```go
+	Taints []Taint  `json:"taints,omitempty"`
+```
+
+to both `NodeSpec` and `NodeStatus`. The value in `NodeStatus` is the union
+of the taints specified by various sources. For now, the only source is
+the `NodeSpec` itself, but in the future one could imagine a node inheriting
+taints from pods (if we were to allow taints to be attached to pods),
+the node's startup coniguration, etc. The scheduler should look at the `Taints`
+in `NodeStatus`, not in `NodeSpec`.
+
```

Comment:
For OpenShift, we allow the concept of a taint to be associated with a namespace (we called it a projectNodeSelector)